iothub/service/src/Messaging/FileUploadNotificationProcessorClient.cs

```diff
@@ -0,0 +1,174 @@
+            }
+        }
+
+        /// <summary>
+        /// Close the AmqpFileNotificationReceiver instance.
+        /// </summary>
+        public virtual async Task CloseAsync()
+        {
+            if (Logging.IsEnabled)
+                Logging.Enter(this, $"Closing AmqpFileNotificationReceiver", nameof(CloseAsync));
+
+            try
+            {
+                await FileNotificationReceiver.CloseAsync().ConfigureAwait(false);
+                await _connection.CloseAsync().ConfigureAwait(false);
+            }
+            catch (Exception ex)
+            {
+                if (Logging.IsEnabled)
+                    Logging.Error(this, $"{nameof(CloseAsync)} threw an exception: {ex}", nameof(CloseAsync));
+                if (ex is IotHubException || ex is IOException)
+                {
+                    _errorProcessor?.Invoke(ex);
+                }
+                throw;
+            }
+            finally
+            {
+                if (Logging.IsEnabled)
+                    Logging.Exit(this, $"Closing AmqpFileNotificationReceiver", nameof(CloseAsync));
+            }
+        }
+
+        /// <inheritdoc />
+        public void Dispose()
+        {
+            if (Logging.IsEnabled)
+                Logging.Enter(this, $"Disposing AmqpFileNotificationReceiver", nameof(Dispose));
+
+            FileNotificationReceiver.Dispose();
+
+            if (Logging.IsEnabled)
+                Logging.Exit(this, $"Disposing AmqpFileNotificationReceiver", nameof(Dispose));
+            GC.SuppressFinalize(this);
+        }
+
+        private async void OnNotificationMessageReceivedAsync(AmqpMessage amqpMessage)
+        {
+            if (Logging.IsEnabled)
+                Logging.Enter(this, amqpMessage, nameof(OnNotificationMessageReceivedAsync));
+
+            try
+            {
+                if (amqpMessage != null && _fileNotificationProcessor != null)
+                {
+                    using (amqpMessage)
+                    {
+                        AmqpClientHelper.ValidateContentType(amqpMessage, CommonConstants.FileNotificationContentType);
+
+                        FileNotification fileNotification = await AmqpClientHelper.GetObjectFromAmqpMessageAsync<FileNotification>(amqpMessage).ConfigureAwait(false);
+                        fileNotification.LockToken = new Guid(amqpMessage.DeliveryTag.Array).ToString();
+
+                        _fileNotificationProcessor.Invoke(fileNotification);
+                    }
+                }
+            }
+            catch(Exception ex)
+            {
+                if (Logging.IsEnabled)
+                    Logging.Error(this, $"{nameof(OnNotificationMessageReceivedAsync)} threw an exception: {ex}", nameof(OnNotificationMessageReceivedAsync));
+                if (ex is IotHubException || ex is IOException)
+                {
+                    _errorProcessor?.Invoke(ex);
```

Comment:
This is a good example of a case where we should invoke the error processor but should **not** throw. This is a callback function that executes on the AMQP layer's thread, not the user's application thread.

Reply:
So keep this line, but don't throw this error below.